### PR TITLE
Standalone gateway package

### PR DESCRIPTION
**Pull request by ssetin**
This feature makes developers able to generate gateway package separately from protobuf .pb files and imports them as outer packages.
It's useful when you have one inner protocol but want to share different api/gateways for different clients

### CODE REVIEW

File: protoc-gen-grpc-gateway/main.go [hunk 1]
```diff
@@ -34,6 +34,7 @@ var (
 	repeatedPathParamSeparator = flag.String("repeated_path_param_separator", "csv", "configures how repeated fields should be split. Allowed values are `csv`, `pipes`, `ssv` and `tsv`.")
 	allowPatchFeature          = flag.Bool("allow_patch_feature", true, "determines whether to use PATCH feature involving update masks (using google.protobuf.FieldMask).")
 	allowColonFinalSegments    = flag.Bool("allow_colon_final_segments", false, "determines whether colons are permitted in the final segment of a path")
+	standalone                 = flag.Bool("standalone", false, "generates standalone gateway package, which imports target service")
```
Comment (johanbrandhorst) [inline protoc-gen-grpc-gateway/main.go:37]:
```suggestion
	standalone                 = flag.Bool("standalone", false, "generates a standalone gateway package, which imports the target service package")
```

Comment (ssetin) [reply protoc-gen-grpc-gateway/main.go:37]: Sure, thank you for the corrections. I will take into account the comments and prepare changes

**Comment by johanbrandhorst** [comment]
Hi Setin, I don't think that merge worked out the way you wanted it to. Maybe you should reopen this PR against v2 from the start? I could try and fix it if you want otherwise but it might be easiest.

**Comment by ssetin** [comment]
Hello Johan, my mistake. I've made rebase to v2 now. I got little problem importing packages in examples because it's placed in the internal folder. I'm solving it

File: protoc-gen-grpc-gateway/main.go (hunk 1 above)
Comment (ssetin) [reply protoc-gen-grpc-gateway/main.go]: I've added notes about the new functionality in readme.md, made generation of examples with the standalone flag set and made rebase to the v2 branch.

File: README.md [hunk 2]
```diff
@@ -160,6 +160,22 @@ annotation to your .proto file
...
+   To set import path of the used stub package you can set its full path in the go_package option
+   or use import_prefix flag.
+   
+   ```sh
+   protoc -I. --grpc-gateway_out=logtostderr=true,grpc_api_configuration=apiOne.yaml,paths=source_relative,standalone=true:gatewayOne \
+     your/service/v1/your_service.proto
+
+   protoc -I. --grpc-gateway_out=logtostderr=true,grpc_api_configuration=apiTwo.yaml,paths=source_relative,standalone=true:gatewayTwo \
+     your/service/v1/your_service.proto
+   ```
```
Comment (johanbrandhorst) [inline README.md:178]:
```suggestion
   
   OR generate a standalone reverse-proxy if needed.

   Suppose you have a generated gRPC stub package, and you want to deploy several
   API gateways using client-specific
   [YAML annotations](https://grpc-ecosystem.github.io/grpc-gateway/docs/grpcapiconfiguration.html).
   You can generate a grpc-gateway which imports the stub as an external 
   package, so you don't have to regenerate it several times.
   To set the import path of the stub package, set its full path in the `go_package`.
   
   ```sh
   protoc -I. --grpc-gateway_out=logtostderr=true,grpc_api_configuration=apiOne.yaml,paths=source_relative,standalone=true:./gen/go/client_one \
     your/service/v1/your_service.proto

   protoc -I. --grpc-gateway_out=logtostderr=true,grpc_api_configuration=apiTwo.yaml,paths=source_relative,standalone=true:./gen/go/client_two \
     your/service/v1/your_service.proto
   ```
```

File: Makefile [hunk 3]
```diff
@@ -173,6 +173,7 @@ $(RUNTIME_TEST_SRCS): $(GO_PLUGIN) $(RUNTIME_TEST_PROTO)
 $(EXAMPLE_GWSRCS): ADDITIONAL_GW_FLAGS:=$(ADDITIONAL_GW_FLAGS),grpc_api_configuration=examples/internal/proto/examplepb/unannotated_echo_service.yaml
 $(EXAMPLE_GWSRCS): $(GATEWAY_PLUGIN) $(EXAMPLES)
 	protoc -I $(PROTOC_INC_PATH) -I. -I$(GOOGLEAPIS_DIR) --plugin=$(GATEWAY_PLUGIN) --grpc-gateway_out=logtostderr=true,allow_repeated_fields_in_body=true,$(PKGMAP)$(ADDITIONAL_GW_FLAGS):. $(EXAMPLES)
+	protoc -I $(PROTOC_INC_PATH) -I. -I$(GOOGLEAPIS_DIR) --plugin=$(GATEWAY_PLUGIN) --grpc-gateway_out=logtostderr=true,allow_repeated_fields_in_body=true,paths=source_relative,import_prefix="github.com/grpc-ecosystem/grpc-gateway/v2/",standalone=true$(ADDITIONAL_GW_FLAGS):examples/internal/proto/standalone $(EXAMPLES)
```
Comment (johanbrandhorst) [inline Makefile:176]: This generates way too many files :joy:. Lets just choose one of the packages (a simple one) and generate a standalone package for that instead of all of them.

Comment (johanbrandhorst) [reply Makefile:176]: What does the `import_prefix` parameter do here? I've never had to use it before.

Comment (ssetin) [reply Makefile]: My point to generate all examples was to check that my feature accept all edge cases described in .proto. Looks like it does, so ok, i leave only one example.
The way you can pass external package path is to set it inside package in go_package option, but I didn't want to change code of the examples, so I decide to simulate such behavior with import_prefix, which adds prefix to go package paths for imported proto files. PS: Sorry for my English.

Comment (johanbrandhorst) [reply Makefile]: This is v2, I'd like to change all of these `go_package`s to be correct anyway. Can we remove `import_prefix` if you update the `go_package`?

**Comment by johanbrandhorst** [comment]
Could we also import the generated package in one of the tests and make sure that it still works?

**Comment by johanbrandhorst** [comment]
https://github.com/grpc-ecosystem/grpc-gateway/blob/master/examples/internal/integration/client_test.go, perhaps.

File: Makefile (hunk 3 above)
Comment (ssetin) [reply Makefile]: no problem

**Comment by ssetin** [comment]
I've added tests for echo service, which handles another api routes. That was tricky to import gw from internal package, so I choose easiest way to get around this issue

File: examples/internal/integration/integration_test.go [hunk 4]
```diff
@@ -43,11 +43,15 @@ func TestEcho(t *testing.T) {
...
-	testEchoBody(t, 8088)
+	for _, apiPrefix := range []string{"v1", "v2"} {
+		t.Run(apiPrefix, func(t *testing.T) {
+			testEcho(t, 8088, apiPrefix, "application/json")
+			testEchoOneof(t, 8088, apiPrefix, "application/json")
+			testEchoOneof1(t, 8088, apiPrefix, "application/json")
+			testEchoOneof2(t, 8088, apiPrefix, "application/json")
+			testEchoBody(t, 8088, apiPrefix)
+		})
+	}
```
Comment (johanbrandhorst) [inline examples/internal/integration/integration_test.go:54]: Why is the apiPrefix necessary?

File: examples/internal/proto/examplepb/unannotated_echo_service.proto [hunk 5]
```diff
@@ -13,12 +13,29 @@ package grpc.gateway.examples.internal.examplepb;
...
 	google.protobuf.Duration duration = 3;
+	oneof code {
+		int64 line_num = 4;
+		string lang = 5;
+	}
+	UnannotatedEmbedded status = 6;
+	oneof ext {
+		int64 en = 7;
+		UnannotatedEmbedded no = 8;
+	}
```
Comment (johanbrandhorst) [inline examples/internal/proto/examplepb/unannotated_echo_service.proto:38]: Why did you change the messages?

Comment (ssetin) [reply examples/internal/proto/examplepb/unannotated_echo_service.proto:38]: Integration tests works with EchoService so I made UnnnnotatedMessage more similar to EchoMesage to reuse all tests.

File: examples/internal/integration/integration_test.go (hunk 4 above)
Comment (ssetin) [reply examples/internal/integration/integration_test.go:54]: I want to test api routes described in echo_service.proto and standalone_echo_service.yaml using same tests

File: Makefile [hunk 6]
```diff
@@ -171,8 +173,12 @@ $(RUNTIME_TEST_SRCS): $(GO_PLUGIN) $(RUNTIME_TEST_PROTO)
...
 
 $(EXAMPLE_GWSRCS): ADDITIONAL_GW_FLAGS:=$(ADDITIONAL_GW_FLAGS),grpc_api_configuration=examples/internal/proto/examplepb/unannotated_echo_service.yaml
+$(EXAMPLE_GWSRCS): ADDITIONAL_SA_FLAGS:=,standalone=true,grpc_api_configuration=examples/internal/proto/examplepb/standalone_echo_service.yaml
 $(EXAMPLE_GWSRCS): $(GATEWAY_PLUGIN) $(EXAMPLES)
-	protoc -I $(PROTOC_INC_PATH) -I. -I$(GOOGLEAPIS_DIR) --plugin=$(GATEWAY_PLUGIN) --grpc-gateway_out=logtostderr=true,allow_repeated_fields_in_body=true,$(PKGMAP)$(ADDITIONAL_GW_FLAGS):. $(EXAMPLES)
+	protoc -I $(PROTOC_INC_PATH) -I. -I$(GOOGLEAPIS_DIR) --plugin=$(GATEWAY_PLUGIN) --grpc-gateway_out=logtostderr=true,allow_repeated_fields_in_body=true,paths=source_relative,$(PKGMAP)$(ADDITIONAL_GW_FLAGS):. $(EXAMPLES)
+	protoc -I $(PROTOC_INC_PATH) -I. -I$(GOOGLEAPIS_DIR) --plugin=$(GATEWAY_PLUGIN) --grpc-gateway_out=logtostderr=true,allow_repeated_fields_in_body=true,paths=source_relative$(ADDITIONAL_SA_FLAGS):examples/internal/proto/standalone $(STANDALONE_EXAMPLES)
+	cp -r examples/internal/proto/standalone/examples/internal/proto/examplepb examples/internal/proto/standalone
+	rm -rf examples/internal/proto/standalone/examples
```
[the remark this answers is not in the document]
Comment (ssetin) [reply Makefile:181]: almost, i'll correct paths a bit

**Comment by johanbrandhorst** [comment]
Great work Setin, thanks for this contribution!
